File: sdk/ocaml/superlog.ml
(*
   superlog.ml - the OCaml client, over Unix sockets and nothing else.

   Copyright 2026 Saxon Herschel Nicholls
   SPDX-License-Identifier: MIT

   Zero dependencies beyond the unix library that ships with the compiler,
   like the Fortran client: a hand-built HTTP POST of NDJSON events. The
   mode comes from SUPERLOG_MODE (development | production) and there is NO
   default, because deciding is the point - unset refuses to start, and
   production sends nothing at all: an inert shell, per the house rule.

     ocamlc -I +unix unix.cma superlog.ml yourapp.ml -o yourapp

   Failures never reach the caller: a logger that can take down the program
   it observes is worse than no logger, so a hub that is down just means
   the next batch counts again.
*)

type t = {
  host : string;
  port : int;
  topic : string;
  app : string;
  device : string;
  session : string;
  mutable seq : int;
  mutable buffer : string list;   (* newest first; reversed at flush *)
  active : bool;
}

let mode () =
  match try Some (Sys.getenv "SUPERLOG_MODE") with Not_found -> None with
  | Some "development" -> `Development
  | Some "production" -> `Production
  | Some other ->
      prerr_endline ("superlog: SUPERLOG_MODE is '" ^ other ^
                     "' - it must be development or production");
      exit 2
  | None ->
      prerr_endline
        "superlog: SUPERLOG_MODE is unset. Declare development or production -\n\
         there is no default, because deciding is the point.";
      exit 2

(* http://host:port with defaults; anything unparseable falls back rather
   than crashing the host program over a URL. *)
let parse_url u =
  let u = try Sys.getenv "SUPER_LOG_URL" with Not_found -> u in
  let stripped =
    if String.length u > 7 && String.sub u 0 7 = "http://"
    then String.sub u 7 (String.length u - 7) else u in
  match String.index_opt stripped ':' with
  | Some i ->
      let host = String.sub stripped 0 i in
      let rest = String.sub stripped (i + 1) (String.length stripped - i - 1) in
      let port = try int_of_string (List.hd (String.split_on_char '/' rest))
                 with _ -> 7333 in
      (host, port)
  | None -> ((if stripped = "" then "127.0.0.1" else stripped), 7333)

let json_escape s =
  let b = Buffer.create (String.length s + 8) in
  String.iter (fun c ->
    match c with
    | '"' -> Buffer.add_string b "\\\""
    | '\\' -> Buffer.add_string b "\\\\"
    | '\n' -> Buffer.add_string b "\\n"
    | '\r' -> Buffer.add_string b "\\r"
    | '\t' -> Buffer.add_string b "\\t"
    | c when Char.code c < 0x20 ->
        Buffer.add_string b (Printf.sprintf "\\u%04x" (Char.code c))
    | c -> Buffer.add_char b c) s;
  Buffer.contents b

let iso_now () =
  let t = Unix.gettimeofday () in
  let tm = Unix.gmtime t in
  Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02d.%03dZ"
    (tm.Unix.tm_year + 1900) (tm.Unix.tm_mon + 1) tm.Unix.tm_mday
    tm.Unix.tm_hour tm.Unix.tm_min tm.Unix.tm_sec
    (int_of_float ((t -. Float.of_int (int_of_float t)) *. 1000.))

let create ?(url = "http://127.0.0.1:7333") ~topic ~app () =
  let active = match mode () with `Development -> true | `Production -> false in
  let host, port = parse_url url in
  {
    host; port; topic; app;
    device = (try Unix.gethostname () with _ -> "ocaml");
    session = Printf.sprintf "%08x" (Random.self_init (); Random.bits () land 0xffffffff);
    seq = 0; buffer = []; active;
  }

(* One TCP connection per flush: a clock at one line a second batches into
   a handful of POSTs a minute, and holding a socket open buys nothing at
   that rate. Errors are swallowed whole - see the header. *)
let flush t =
  if t.active && t.buffer <> [] then begin
    let body = String.concat "\n" (List.rev t.buffer) in
    t.buffer <- [];
    try
      let addr = Unix.ADDR_INET ((Unix.gethostbyname t.host).Unix.h_addr_list.(0), t.port) in
      let sock = Unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
      (try
        Unix.connect sock addr;
        let req = Printf.sprintf
          "POST /ingest/%s HTTP/1.1\r\nHost: %s\r\nContent-Type: application/x-ndjson\r\n\
           Content-Length: %d\r\nConnection: close\r\n\r\n%s"
          t.topic t.host (String.length body) body in
        ignore (Unix.write_substring sock req 0 (String.length req));
        (* Read and discard the reply so the hub never sees a reset mid-answer. *)
        let buf = Bytes.create 256 in
        ignore (try Unix.read sock buf 0 256 with _ -> 0)
      with _ -> ());
      Unix.close sock
    with _ -> ()                    (* hub down; the next batch counts again *)
  end

let log ?(fields = []) ?metric t level msg =
  if t.active then begin
    let field_json =
      match fields with
      | [] -> ""
      | fs ->
          ",\"fields\":{" ^
          String.concat ","
            (List.map (fun (k, v) ->
               "\"" ^ json_escape k ^ "\":\"" ^ json_escape v ^ "\"") fs) ^ "}"
    in
    let metric_json =
      match metric with
      | None -> ""
      | Some (name, value) ->
          Printf.sprintf ",\"metric\":{\"name\":\"%s\",\"value\":%g}"
            (json_escape name) value
    in
    let line = Printf.sprintf
      "{\"v\":1,\"ts\":\"%s\",\"seq\":%d,\"session\":\"%s\",\"level\":\"%s\",\
       \"origin\":{\"runtime\":\"ocaml\",\"app\":\"%s\",\"platform\":\"host\",\"device\":\"%s\"},\
       \"tag\":\"%s\",\"msg\":\"%s\"%s%s}"
      (iso_now ()) t.seq t.session level (json_escape t.app)
      (json_escape t.device) (json_escape t.app) (json_escape msg)
      metric_json field_json in
    t.seq <- t.seq + 1;
    t.buffer <- line :: t.buffer;
    if List.length t.buffer >= 16 then flush t
  end

let debug ?fields t msg = log ?fields t "DEBUG" msg
let info ?fields t msg = log ?fields t "INFO" msg
let warn ?fields t msg = log ?fields t "WARN" msg
let error ?fields t msg = log ?fields t "ERROR" msg
let metric t name value =
  log t "DEBUG" (Printf.sprintf "%s =%g" name value) ~metric:(name, value)

File: sdk/haskell/SuperLog.hs
{-# LANGUAGE CPP #-}
--
--  SuperLog.hs - the Haskell client: GHC's boot libraries, plus curl.
--
--  Copyright 2026 Saxon Herschel Nicholls
--  SPDX-License-Identifier: MIT
--
--  base has no sockets and this repo takes no dependencies, so the wire is
--  curl - the same honest bargain the shell producer makes, and curl is on
--  every machine this bench will meet. Everything else is boot libraries
--  that ship inside GHC (process, time).
--
--  The mode is compile-time, like the C++ and Fortran clients:
--
--    ghc -DDEVELOPMENT ... SuperLog.hs yourapp.hs
--
--  Exactly one of -DDEVELOPMENT / -DPRODUCTION must be given; neither or
--  both refuses to compile, because deciding is the point. PRODUCTION
--  compiles to an inert shell that sends nothing at all.
--

module SuperLog
  ( Log, newLog, flushLog
  , logMsg, debug, info, warn, err, metric
  ) where

#if defined(DEVELOPMENT) && defined(PRODUCTION)
#error "declare DEVELOPMENT xor PRODUCTION, not both - deciding is the point"
#endif
#if !defined(DEVELOPMENT) && !defined(PRODUCTION)
#error "declare -DDEVELOPMENT or -DPRODUCTION - there is no default"
#endif

import Data.Char (isControl, ord)
import Data.IORef
import Data.List (intercalate)
import Data.Time.Clock (getCurrentTime)
import Data.Time.Format (defaultTimeLocale, formatTime)
import System.Environment (lookupEnv)
import System.Process (readProcessWithExitCode)
import Text.Printf (printf)

data Log = Log
  { lgUrl :: String
  , lgTopic :: String
  , lgApp :: String
  , lgDevice :: String
  , lgSession :: String
  , lgSeq :: IORef Int
  , lgBuf :: IORef [String]     -- newest first; reversed at flush
  }

newLog :: String -> String -> IO Log
newLog topic app = do
  url <- maybe "http://127.0.0.1:7333" id <$> lookupEnv "SUPER_LOG_URL"
  -- hostname(1) rather than a networking package: one call at startup.
  (_, hn, _) <- readProcessWithExitCode "hostname" ["-s"] ""
  now <- getCurrentTime
  let device = case lines hn of { (h:_) -> h; _ -> "haskell" }
      -- Picosecond digits are as random as a session id needs to be.
      session = take 8 (formatTime defaultTimeLocale "%q" now ++ "00000000")
  seqRef <- newIORef 0
  buf <- newIORef []
  pure (Log url topic app device session seqRef buf)

jsonEscape :: String -> String
jsonEscape = concatMap esc
  where
    esc '"' = "\\\""
    esc '\\' = "\\\\"
    esc '\n' = "\\n"
    esc '\r' = "\\r"
    esc '\t' = "\\t"
    esc c | isControl c = printf "\\u%04x" (ord c)
          | otherwise = [c]

isoNow :: IO String
isoNow = formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" <$> getCurrentTime

-- | One event; buffered, flushed every 16 or on flushLog. In PRODUCTION
-- this whole body compiles away to pure ().
logMsg :: Log -> String -> String -> [(String, String)] -> IO ()
#if defined(PRODUCTION)
logMsg _ _ _ _ = pure ()
#else
logMsg lg level msg fields = do
  ts <- isoNow
  n <- atomicModifyIORef' (lgSeq lg) (\s -> (s + 1, s))
  let fjson = case fields of
        [] -> ""
        fs -> ",\"fields\":{" ++
              intercalate "," [ "\"" ++ jsonEscape k ++ "\":\"" ++ jsonEscape v ++ "\""
                              | (k, v) <- fs ] ++ "}"
      line = concat
        [ "{\"v\":1,\"ts\":\"", ts, "\",\"seq\":", show n
        , ",\"session\":\"", lgSession lg
        , "\",\"level\":\"", level
        , "\",\"origin\":{\"runtime\":\"haskell\",\"app\":\"", jsonEscape (lgApp lg)
        , "\",\"platform\":\"host\",\"device\":\"", jsonEscape (lgDevice lg)
        , "\"},\"tag\":\"", jsonEscape (lgApp lg)
        , "\",\"msg\":\"", jsonEscape msg, "\"", fjson, "}" ]
  pending <- atomicModifyIORef' (lgBuf lg) (\b -> (line : b, length b + 1))
  if pending >= 16 then flushLog lg else pure ()
#endif

-- | POST the batch; a hub that is down means the next batch counts again,
-- never an exception in the host program.
flushLog :: Log -> IO ()
#if defined(PRODUCTION)
flushLog _ = pure ()
#else
flushLog lg = do
  batch <- atomicModifyIORef' (lgBuf lg) (\b -> ([], reverse b))
  case batch of
    [] -> pure ()
    ls -> do
      _ <- readProcessWithExitCode "curl"
             [ "-s", "-m", "5", "-o", "/dev/null"
             , "-X", "POST", lgUrl lg ++ "/ingest/" ++ lgTopic lg
             , "-H", "content-type: application/x-ndjson"
             , "--data-binary", intercalate "\n" ls ] ""
      pure ()
#endif

debug, info, warn, err :: Log -> String -> [(String, String)] -> IO ()
debug lg m = logMsg lg "DEBUG" m
info lg m = logMsg lg "INFO" m
warn lg m = logMsg lg "WARN" m
err lg m = logMsg lg "ERROR" m

-- | A reading for the chart: DEBUG, with the metric riding the event.
metric :: Log -> String -> Double -> IO ()
#if defined(PRODUCTION)
metric _ _ _ = pure ()
#else
metric lg name value = do
  ts <- isoNow
  n <- atomicModifyIORef' (lgSeq lg) (\s -> (s + 1, s))
  let line = concat
        [ "{\"v\":1,\"ts\":\"", ts, "\",\"seq\":", show n
        , ",\"session\":\"", lgSession lg
        , "\",\"level\":\"DEBUG\",\"origin\":{\"runtime\":\"haskell\",\"app\":\""
        , jsonEscape (lgApp lg), "\",\"platform\":\"host\",\"device\":\""
        , jsonEscape (lgDevice lg), "\"},\"tag\":\"", jsonEscape (lgApp lg)
        , "\",\"msg\":\"", jsonEscape name, " =", show value
        , "\",\"metric\":{\"name\":\"", jsonEscape name
        , "\",\"value\":", show value, "}}" ]
  pending <- atomicModifyIORef' (lgBuf lg) (\b -> (line : b, length b + 1))
  if pending >= 16 then flushLog lg else pure ()
#endif
